pkg/plugins/tainttoleration/taint_toleration.go

```diff
@@ -0,0 +1,171 @@
+package tainttoleration
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"sort"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+	clusterapiv1 "open-cluster-management.io/api/cluster/v1"
+	clusterapiv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+	"open-cluster-management.io/placement/pkg/plugins"
+)
+
+var _ plugins.Filter = &TaintToleration{}
+
+var timeDiffs = sort.Float64Slice{}
+var TolerationClock = (clock.Clock)(clock.RealClock{})
+
+const (
+	placementLabel = "cluster.open-cluster-management.io/placement"
+	description    = "TaintToleration is a plugin that checks if a placement tolerates a managed cluster's taints"
+)
+
+type TaintToleration struct {
+	handle plugins.Handle
+}
+
+func New(handle plugins.Handle) *TaintToleration {
+	return &TaintToleration{
+		handle: handle,
+	}
+}
+
+func (p *TaintToleration) Name() string {
+	return reflect.TypeOf(*p).Name()
+}
+
+func (pl *TaintToleration) Description() string {
+	return description
+}
+
+func (pl *TaintToleration) Filter(ctx context.Context, placement *clusterapiv1beta1.Placement, clusters []*clusterapiv1.ManagedCluster) ([]*clusterapiv1.ManagedCluster, *time.Duration, error) {
+	if len(clusters) == 0 {
+		return clusters, nil, nil
+	}
+
+	// do validation on each toleration and return error if necessary
+	for _, toleration := range placement.Spec.Tolerations {
+		if len(toleration.Key) == 0 && toleration.Operator != clusterapiv1beta1.TolerationOpExists {
+			return nil, nil, errors.New("If the key is empty, operator must be Exists.\n")
+		}
+		if toleration.Operator == clusterapiv1beta1.TolerationOpExists && len(toleration.Value) > 0 {
+			return nil, nil, errors.New("If the operator is Exists, the value should be empty.\n")
+		}
+	}
+
+	existingDecisions := getDecisions(pl.handle, placement)
+
+	matched := []*clusterapiv1.ManagedCluster{}
+	for _, cluster := range clusters {
+		if isClusterTolerated(cluster, placement.Spec.Tolerations, existingDecisions.Has(cluster.Name)) {
+			matched = append(matched, cluster)
+		}
+	}
+
+	//calculate requeue time
+	if len(timeDiffs) > 0 {
+		sort.Float64s(timeDiffs)
+		requeueTime := time.Duration(timeDiffs[0])
+		return matched, &requeueTime, nil
+	}
+
+	return matched, nil, nil
+}
+
+// isClusterTolerated returns true if a cluster is tolerated by the given toleration array
+func isClusterTolerated(cluster *clusterapiv1.ManagedCluster, tolerations []clusterapiv1beta1.Toleration, exist bool) bool {
+	for _, taint := range cluster.Spec.Taints {
+		if !isTaintTolerated(taint, tolerations, exist) {
+			return false
+		}
+	}
+	return true
+}
+
+// isTaintTolerated returns true if a taint is tolerated by the given toleration array
+func isTaintTolerated(taint clusterapiv1.Taint, tolerations []clusterapiv1beta1.Toleration, exist bool) bool {
+	for _, toleration := range tolerations {
```

Comment:
One question for the design, suppose one cluster is already in the decision, then the cluster adds a taint with TaintEffectNoSelectIfNew and placement tolerations are empty, should we keep it in the decision?